mdpow/analysis/ensemble.py

```diff
@@ -516,16 +516,17 @@ def run(self, start=None, stop=None, step=None):
         self._prepare_ensemble()
         for self._key in ProgressBar(self._ensemble.keys(), verbose=True):
             self._setup_system(self._key, start=start, stop=stop, step=step)
-            self._prepare_universe()
-            self._single_universe()
-            for i, ts in enumerate(ProgressBar(self._trajectory[self.start:self.stop:self.step], verbose=True,
+            try:
+                self._single_universe()
+            except NotImplementedError:
```

Comment:
Need to document that NotImplementedError is a sign to do single frame analysis.

Reply:
Unfortunately I think this method will not work based on how the EnsembleAnalysis base class is written in ensemble.py or how inheritance is being passed to DihedralAnalysis. 

`    def _single_universe(self):
        """Calculations on a single Universe object.
            Run on each universe in the ensemble during when
            self.run in called.
        """
        pass  # pragma: no cover

    def _single_frame(self):
        """Calculate data from a single frame of trajectory
        Called on each frame for universes in the Ensemble.
        """
        pass  # pragma: no cover`

This allowed for both self._single_universe() and self._single_frame() to be called in the same block every  time .run() was called by an EnsembleAnalysis subclass (pre try/except), which means NotImplementedError is never being raised even when _single_universe() is only defined in the base class and not the subclass. 

For example, with the try/except method as it is here now, when running DihedralAnalysis, self._single_universe() is called, nothing happens because it is only defined with 'pass' in the base class, NotImplementedError is not raised, and the except block is not executed. Therefore, self._single_frame() is never called, and no data is analyzed or stored in the results DataFrame expected from DihedralAnalysis. I tested this on solvation shell analysis, which also uses _single_frame(), and got the same problem, an empty results DataFrame.

I am looking into alternative options that don't require modifying anything outside of the run method.

Reply:
Modify the base class and `raise NotImplementedError` instead of `pass`. This will ensure that developers implement at least one of the two.